### PR TITLE
Minion segment conversion framework

### PR DESCRIPTION
**Pull request by snleee**
1. Added interface for record partitioner, transformer, aggregator
2. Added mapper and reduce record reader
3. Added segment converter that puts all the logics together

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/minion/SegmentConverter.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import com.linkedin.pinot.core.segment.creator.impl.SegmentIndexCreationDriverImpl;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+
+/**
+ * Genenric Segment Converter
```
[the remark this answers is not in the document]
Comment (snleee) [reply pinot-core/src/main/java/com/linkedin/pinot/core/minion/SegmentConverter.java]: fixed

File: pinot-core/src/main/java/com/linkedin/pinot/core/minion/segment/DefaultRecordPartitioner.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Default record partitioner that is using a hash code of the input row
+ */
+public class DefaultRecordPartitioner implements RecordPartitioner {
+
+  public DefaultRecordPartitioner(){
+  }
+
+  @Override
+  public int getPartitionFromRecord(GenericRow row, int numPartition) {
```
[the remark this answers is not in the document]
Comment (snleee) [reply pinot-core/src/main/java/com/linkedin/pinot/core/minion/segment/DefaultRecordPartitioner.java]: fixed

File: pinot-core/src/main/java/com/linkedin/pinot/core/minion/segment/RecordAggregator.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ */
+public interface RecordAggregator {
+
+  /**
+   * Aggregate grouped records into a single row
+   *
+   * @param rows a group of rows that shares the same values for the aggregation columns
+   * @return an aggregated row
+   */
+  GenericRow aggregateRecords(List<GenericRow> rows);
```
[the remark this answers is not in the document]
Comment (snleee) [reply pinot-core/src/main/java/com/linkedin/pinot/core/minion/segment/RecordAggregator.java]: I would want to enforce to contain all columns. I will add this to documentation.

File: pinot-core/src/main/java/com/linkedin/pinot/core/minion/SegmentConverter.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  private RecordTransformer _recordTransformer;
+
+  // Optional
+  private int _totalNumPartition;
+  private RecordPartitioner _recordPartitioner;
+  private RecordAggregator _recordAggregator;
+  private List<String> _aggregationColumns;
+  private IndexingConfig _indexingConfig;
+
+  public SegmentConverter(@Nonnull List<File> inputIndexDirs, @Nonnull File workingDir, @Nonnull String tableName,
```
[the remark this answers is not in the document]
Comment (snleee) [reply pinot-core/src/main/java/com/linkedin/pinot/core/minion/SegmentConverter.java]: fixed

File: pinot-core/src/main/java/com/linkedin/pinot/core/minion/SegmentConverter.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    }
+    return resultFiles;
+  }
+
+  /**
+   * Helper function to trigger the segment creation
+   */
+  private void buildSegment(String outputPath, String tableName, String segmentName, RecordReader recordReader,
+      IndexingConfig indexingConfig) throws Exception {
+    SegmentGeneratorConfig segmentGeneratorConfig = new SegmentGeneratorConfig(recordReader.getSchema());
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/minion/SegmentConverter.java]: This is not reading everything, so you would miss things like RAW index etc. This requires looking at segment metadata to derive all such info.

File: pinot-core/src/main/java/com/linkedin/pinot/core/minion/segment/DefaultRecordPartitioner.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.linkedin.pinot.core.data.GenericRow;
+
+
+/**
+ * Default record partitioner that is using a hash code of the input row
+ */
+public class DefaultRecordPartitioner implements RecordPartitioner {
+
+  public DefaultRecordPartitioner(){
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/minion/segment/DefaultRecordPartitioner.java]: Style check.

File: pinot-core/src/main/java/com/linkedin/pinot/core/minion/segment/ReducerRecordReader.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    if (!_nextRowReturned) {
+      return true;
+    }
+
+    while(_recordReader.hasNext()) {
+      GenericRow currentRow = _recordReader.next();
+
+      // Grouping rows by the given aggregation columns
+      if (_rowGroup.isEmpty() || haveSameAggregationColumns(_rowGroup.iterator().next(), currentRow)) {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/minion/segment/ReducerRecordReader.java]: It is GroupByColumns, not AggregationColumns, right?

**Pull request merged**